simba/schedule.py

```diff
@@ -285,10 +301,170 @@ def assign_vehicles(self):
 
         self.vehicle_type_counts = vehicle_type_counts
 
+    def assign_vehicles_w_adaptive_soc(self, args):
+        """ Assign vehicle IDs to rotations.
+
+        A greedy approach is used.
+        For every rotation it is checked if an existing vehicle of the same type at the same
+        location has enough energy to service the rotation, not accounting for possible opportunity
+        charging of the next rotation. If multiple such vehicles exist, the one with the lowest soc
+        is used. If no vehicle is available, a new vehicle ID is generated.
+        :param args: arguments
+        :type args: Namespace
+        """
+        rotations_in_progress = []
+        all_standing_vehicles = []
+        vehicle_data = dict()
+        initial_soc = args.desired_soc_deps
+
+        # count number of vehicles per type
+        # used for unique vehicle id e.g. vehicletype_chargingtype_id
+        vehicle_type_counts = {f'{vehicle_type}_{charging_type}': 0
+                               for vehicle_type, charging_types in self.vehicle_types.items()
+                               for charging_type in charging_types.keys()}
+
+        charge_levels = {station.get(param) for name, station in self.stations.items()
+                         for param in ["cs_power_deps_oppb", "cs_power_deps_depb"]
+                         if station["type"] == "deps"}
+        # Remove None as charge level
+        charge_levels = charge_levels.difference([None])
+        # Add default values from arguments
+        charge_levels.add(self.cs_power_deps_depb)
+        charge_levels.add(self.cs_power_deps_oppb)
+
+        # Calculates numeric charge curves for each charge_level. The charge levels might clip the
+        # charge curve of the vehicle.
+        charge_curves = self.get_charge_curves(charge_levels, initial_soc, time_step_min=1)
+
+        rotations = sorted(self.rotations.values(), key=lambda rot: rot.departure_time)
+        for k, rot in enumerate(rotations):
+            # find vehicles that have completed their rotation
+            while rotations_in_progress:
+                r = rotations_in_progress.pop(0)
+                if rot.departure_time > r.arrival_time:
+                    all_standing_vehicles.append((r.vehicle_id, r.arrival_name))
+                else:
+                    rotations_in_progress.insert(0, r)
+                    break
+
+            # find standing vehicle for rotation if exists
+            # else generate new vehicle id
+            vt = rot.vehicle_type
+            ct = rot.charging_type
+            vt_ct = f"{vt}_{ct}"
+
+            station_is_electrified = True
+            try:
+                self.stations[rot.departure_name]
+            except KeyError:
+                logging.warning(f"Rotation {rot.id} ends at a non electrified station.")
+                station_is_electrified = False
+
+            # filter vehicles of the right vehicle type and right location
```

Comment:
what is the "right" vt/location? => get all vehicles of same vehicle type at departure station